File: castvibe/_models/_connection.py
"""Connection namespace messages (CONNECT / CLOSE)."""

from typing import Annotated, Any, Literal

from pydantic import Discriminator, TypeAdapter

from castvibe._models._base import CastModel


class SenderInfo(CastModel):
    """Metadata about the connecting sender (embedded in CONNECT)."""

    sdk_type: int | None = None
    version: str | None = None
    browser_version: str | None = None
    platform: int | None = None
    connection_type: int | None = None
    model: str | None = None
    system_version: str | None = None


class ConnectRequest(CastModel):
    """Virtual connection request from a sender.

    The ``conn_type`` field is present in some sender implementations
    (observed as a numeric value in go-cast).  We accept it but don't
    require it.
    """

    type: Literal["CONNECT"] = "CONNECT"
    origin: dict[str, Any] = {}
    user_agent: str | None = None
    sender_info: SenderInfo | None = None
    conn_type: int | None = None


class CloseRequest(CastModel):
    """Virtual connection close from a sender."""

    type: Literal["CLOSE"] = "CLOSE"
    reason_code: int | None = None


ConnectionMessage = Annotated[
    ConnectRequest | CloseRequest,
    Discriminator("type"),
]
"""Discriminated union of all connection namespace messages."""

connection_message_adapter: TypeAdapter[ConnectionMessage] = TypeAdapter(
    ConnectionMessage,
)

File: crates/vibecast-messages/src/connection.rs
//! Connection namespace messages (CONNECT / CLOSE).

use serde::Deserialize;
use serde_json::Value;

/// Metadata about the connecting sender (embedded in CONNECT).
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SenderInfo {
    /// SDK type.
    pub sdk_type: Option<i64>,
    /// SDK version.
    pub version: Option<String>,
    /// Browser version.
    pub browser_version: Option<String>,
    /// Platform code.
    pub platform: Option<i64>,
    /// Connection type code.
    pub connection_type: Option<i64>,
    /// Sender model.
    pub model: Option<String>,
    /// System version.
    pub system_version: Option<String>,
}

/// Virtual connection request from a sender.
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectRequest {
    /// Origin object (opaque).
    #[serde(default)]
    pub origin: Value,
    /// Sender user agent.
    pub user_agent: Option<String>,
    /// Sender metadata.
    pub sender_info: Option<SenderInfo>,
    /// Connection type (numeric in some senders).
    pub conn_type: Option<i64>,
}

/// Virtual connection close from a sender.
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CloseRequest {
    /// Reason code.
    pub reason_code: Option<i64>,
}

/// Discriminated union of connection namespace messages.
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type")]
pub enum ConnectionMessage {
    /// CONNECT.
    #[serde(rename = "CONNECT")]
    Connect(ConnectRequest),
    /// CLOSE.
    #[serde(rename = "CLOSE")]
    Close(CloseRequest),
}
